File: programmers/[programmers] 이모티콘 할인행사.js
function solution(users, emoticons) {
  let amounts = new Array(users.length).fill(0);
  let max_members = 0,
    max_amounts = 0;
  const dfs = (emoticonIndex) => {
    if (emoticonIndex >= emoticons.length) {
      let members = 0,
        total = 0;
      for (let i = 0; i < users.length; i++) {
        if (users[i][1] <= amounts[i]) members += 1;
        else total += amounts[i];
      }
      if (
        members > max_members ||
        (members === max_members && total > max_amounts)
      ) {
        (max_members = members), (max_amounts = total);
      }
      return;
    }

    for (const discount of [10, 20, 30, 40]) {
      const tmpAmounts = [...amounts];
      for (let i = 0; i < users.length; i++) {
        if (users[i][0] <= discount)
          amounts[i] += (emoticons[emoticonIndex] * (100 - discount)) / 100;
      }
      dfs(emoticonIndex + 1);
      amounts = [...tmpAmounts];
    }
  };
  dfs(0);
  return [max_members, max_amounts];
}

File: programmers/[programmers] 이모티콘 할인행사.py
def solution(users, emoticons):
    amounts = [0]*len(users)
    max_amount = max_member = 0

    def dfs(emoticon_index):
        nonlocal amounts, max_amount, max_member
        if emoticon_index >= len(emoticons):
            members = total = 0
            for i in range(len(users)):
                if users[i][1] <= amounts[i]: members+=1
                else: total += amounts[i]
            if members > max_member or (members == max_member and total > max_amount):
                max_member = members
                max_amount = total
            return
        for discount in [10, 20, 30, 40]:
            tmp_amount = amounts[:]
            for i in range(len(users)):
                if users[i][0] <= discount:
                    amounts[i] += (emoticons[emoticon_index]*(100-discount)/100)
            dfs(emoticon_index+1)
            amounts = tmp_amount[:]
        return

    dfs(0)
    return [max_member, max_amount]